core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -3075,6 +3086,7 @@ private InFlightState tryUpdateState(RecordState newState, DeliveryCountOps ops,
                 return this;
             } catch (IllegalStateException e) {
                 log.error("Failed to update state of the records", e);
+                rollbackState = null;
```

Comment:
I'm not certain that `rollbackState` is initialized in all situations. For example, if we come through `acquireSubsetBatchRecords` into `tryUpdateState`, then wouldn't `rollbackState` be null and therefore the transition doesn't appear to be in progress. Maybe I'm reading this incorrectly, but I think there's a gap.

Reply:
You are right, there is no rollbackState in case `tryUpdateState` is called directly without starting a transition. The rollback is required for the cases when the state change is not certain i.e. a persisiter call is required and correspoding response determines whether state should be committed or not. In case of acquire we do update directly as no persister call is needed.